dev_tools/modules.py

```diff
@@ -0,0 +1,166 @@
+) -> List[Module]:
+    """Returns a list of python modules based defined by setup.py files.
+
+    Args:
+        include_parent: if true, a setup.py is expected in `search_dir`, and the corresponding
+        module will be included.
+        search_dir: the search directory for modules, by default the repo root.
+    Returns:
+          a list of `Module`s that were found, where each module `m` is initialized with `m.root`
+           relative to `search_dir`, `m.raw_setup` contains the dictionary equivalent to the
+           keyword args passed to the `setuptools.setup` method in setup.py
+    """
+
+    relative_folders = sorted(
+        f.relative_to(search_dir)
+        for f in search_dir.glob("*")
+        if f.is_dir() and (f / "setup.py").is_file()
+    )
+    if include_parent:
+        parent_setup_py = search_dir / "setup.py"
+        assert parent_setup_py.exists(), (
+            f"include_parent=True, but {parent_setup_py} " f"does not exist."
+        )
+        relative_folders.append(Path("."))
+
+    result = [
+        Module(root=folder, raw_setup=_parse_module(search_dir / folder))
+        for folder in relative_folders
+    ]
+
+    return result
+
+
+def _parse_module(folder: Path) -> Dict[str, Any]:
+    setup_args = {}
+    import setuptools
+
+    orig_setup = setuptools.setup
+    cwd = os.getcwd()
+
+    def setup(**kwargs):
+        setup_args.update(kwargs)
+
+    try:
+        setuptools.setup = setup
+        os.chdir(str(folder))
+        setup_py = open("setup.py").read()
+        exec(setup_py, globals())
+        assert setup_args, f"Invalid setup.py - setup() was not called in {folder}/setup.py!"
+        return setup_args
+    except BaseException:
+        print(f"Failed to run {folder}/setup.py:")
+        raise
+    finally:
+        setuptools.setup = orig_setup
+        os.chdir(cwd)
+
+
+def _print_list_modules(mode: str, include_parent: bool = False):
+    """Prints certain properties of cirq modules on separate lines.
+
+    Module root folder and top level package paths are supported. The search dir is the current
+    directory.
+
+    Args:
+        mode: 'folder' lists the root folder for each module, 'package-path' lists the path to
+         the top level package(s).
+        include_cirq: when true the cirq metapackage is included in the list
+    Returns:
+          a list of strings
+    """
+    for m in list_modules(Path("."), include_parent):
+        if mode == _FOLDER:
+            print(m.root, end=" ")
+        elif mode == _PACKAGE_PATH:
+            for p in m.top_level_package_paths:
+                print(p, end=" ")
+
+
+def main(argv: List[str]):
+    args = parse(argv)
+    # args.func is where we store the function to be called for a given subparser
+    # e.g. it is list_modules for the `list` subcommand
+    f = args.func
+    # however the func is not going to be needed for the function itself, so
+    # we remove it here
+    delattr(args, 'func')
```

Comment:
nit: could do `del args.func` since we know the attribute name statically. Up to you.